### PR TITLE
Fix "Stairlock" Exploit

### PR DESCRIPTION
**Pull request by kphoenix137**
This fix prevents monsters from pathing into a trigger tile in the Cathedral and Crypt levels, fixing the infamous "Butcher Stairlock Bug". Monsters can still occupy these tiles (example: Player uses knockback item), however they will not willingly walk into the tile in question.

### CODE REVIEW

File: Source/monster.cpp [hunk 1]
```diff
@@ -3043,6 +3043,15 @@ bool IsRelativeMoveOK(const Monster &monster, Point position, Direction mdir)
 	Point futurePosition = position + mdir;
 	if (!InDungeonBounds(futurePosition) || !IsTileAvailable(monster, futurePosition))
 		return false;
+	// Fixes exploit that allows a monster to become "stuck" in a trigger tile in Cathedral and Crypt levels
+	if ((currlevel >= 1 && currlevel <= 4)
+	    || (currlevel >= 21 && currlevel <= 24)) {
```
[the remark this answers is not in the document]
Comment (kphoenix137) [reply Source/monster.cpp:3048]: VS did it automatically so I just assumed is was "correct formatting" xD

**Comment by AJenbo** [comment]
Won't this also makde them avoid town portals and possibly other random things? Maybe we should think about looking in to why they do not path out of that spot when they end up being stuck.

**Comment by StephenCWills** [comment]
It's because `RandomWalk()` does not allow him to move away from his target.

https://github.com/diasurgical/devilutionX/blob/25a2b59f89c922e90bf722d7940af272e7d4df4c/Source/monster.cpp#L1642-L1663

1. Try to walk directly toward the target.
2. If that doesn't work, try turning 45 degrees in both directions.
3. If neither have worked, try turning 90 degrees in both directions.
4. If none of those worked, give up and stand in the stairs I guess.

I suppose it's possible that you might be able to get him stuck if a sarcophagus spawns next to an arch, but I believe that's pretty rare or maybe not even possible. There is no dungeon formation more reliable than the stairs for this trick.

**Comment by AJenbo** [comment]
It's pretty impressive that a strategy that simple works so well in general.

Maybe a better solution is to add some blacklisted tile ids to `IsTileSafe()`. I think it stands to reason for a monster not to feel safe standing in that position.

One down side about both methods is that you can push a monster past the trigger of Leoric's and Hell's down stairs and they would get suck behind it. A workaround for that would be to not consider `IsTileSafe()` if they are already standing on an unsafe tile, which should also be a general improvement.

**Pull request closed without merging**